lib/io/hardware_gpio.cpp

```diff
@@ -19,18 +39,58 @@ HardwareGpio::HardwareGpio(hyped::core::ILogger &log) : log_(log)
   // TODO: implement
 }
 
+
+
 std::optional<std::shared_ptr<IGpioReader>> HardwareGpio::getReader(const uint8_t pin)
 {
-  // TODO: implement
-  log_.log(hyped::core::LogLevel::kFatal, "GPIO reader not implemented");
-  return std::nullopt;
+  //What happens if all shared pointers are removed and this is called?
+  //Map solves this, GPIO will always hold one pointer so it dosen't get destroyed.
+  if (InitializedReaders.count(pin) != 0) {
+    std::shared_ptr<HardwareGpioReader> reader;
+    return reader;
+  }
+  const uint8_t bank = pin / 32;
```

Comment:
is `pin` definitely divisible by 32? Does this definitely behave as expected?

Reply:
This was taken from the old code, we're purposefully doing integer division here (ignore remainder), to get the bank numbers of 0, 1, 2, 3 which then we can get the addresses from bankAddresses.

Reply:
It being in old code does not make it right. Last year's code had lots of problems and just copying code without verifying it is how problems get carried over.